spritecontrol.cpp

```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+
+		static std::map<Id, Point> velocity;
+		static std::map<Id, double> desired_velocity;
+		static std::map<Id, double> desired_separation;
+		static std::map<Id, double> vision_range;
+		static std::map<Id, double> blind_angle;
+
+		// Gives a random value between (1 - variation; 1 + variation).
+		// Exponent affects the bias of the curve towards 1 before rapidly diverging at the edges.
+		// Slope affects how linear the curve is. Slope = 1 behaves like exponent = 1.
+		// Variation shouldn't be larger than 1. Exponent must be larger than 0. Slope should be from 0 to 1.
+		auto random_curve = [](double variation, double exponent = 5.0, double slope = 0.1) {
+			double random = Noise::random();
+			double sign = Noise::random() < 0.5 ? -1.0 : 1.0;
+			random = random * slope + pow(random, exponent) * (1 - slope);
+			return 1.0 + sign * random * variation;		// Equation: 1 + sign(rand) * (|rand| * slope + |rand|^exp * (1 - slope)) * variation
+		};
+
+		// Converts a vector to an angle in the range (-M_PI, M_PI].
+		auto get_angle = [](const Point &vector) {
+			double x = get<X>(vector);
+			double y = get<Y>(vector);
+
+			if (x == 0 && y == 0) {
+				return 0.0;
+			}
+
+			double angle = atan2(y, x);
+
+			// i'm sure there will be absolutely nothing wrong with this
+			if (x >= 0) {
+				return angle;
+			} else if (y >= 0) {
+				return angle + M_PI;
+			} else {
+				return angle - M_PI;
+			}
+		};
+
+		// Wraps angles to be inside the range (-M_PI, M_PI]
+		auto wrap_angle = [](double angle) {
+			angle += M_PI;
+			angle -= 2 * M_PI * floor(angle / (2 * M_PI));
+			return angle - 2 * M_PI;
+		};
+
+		auto get_vector_magnitude = [](const Point &vector) {
+			double x = get<X>(vector);
+			double y = get<Y>(vector);
+
+			return sqrt(x * x + y * y);
+		};
+
+		auto normalize_vector = [get_vector_magnitude](const Point &vector) {
+			double magnitude = get_vector_magnitude(vector);
+			return magnitude != 0.0 ? vector / magnitude : vector;
+		};
+
+		if (velocity.empty() || desired_velocity.empty() || desired_separation.empty() || vision_range.empty() || blind_angle.empty()) {
+			for (const Sprite *sprite : *sprites) {
+				double magnitude = Noise::random() + 0.4;
+				double radians = Noise::random() * M_PI * 2;
+				velocity[sprite->id()] = Point(std::cos(radians) * magnitude, std::sin(radians) * magnitude);
+				desired_velocity[sprite->id()] = magnitude;
+
+				double separation_radius = random_curve(0.2) * SEPARATION_X_RADIUS;
+				desired_separation[sprite->id()] = separation_radius;
+
+				double vision_radius = random_curve(0.5) * VISION_X_RADIUS;
+				vision_range[sprite->id()] = vision_radius;
+
+				double degrees = random_curve(1.0, 10.0, 0.25) * BLIND_DEGREES;
+				blind_angle[sprite->id()] = degrees * M_PI / 180.0;
+			}
+		}
+
+		std::map<Sprite *, Point> separation_velocity_changes;
```

Comment:
These maps are probably hurting the performance the most. I don't have any super specific optimization plan but I think these eventually turn into static arrays, OR maybe disappear altogether (can we just apply the transformations now instead of storing them and applying them later?)

Reply:
I imagine these could probably be static arrays that's the same size as the sprites vector.
The main reason I'm not applying the transformations here is cause that'd mean later sprites read different values than earlier sprites. but whether that's actually an issue I'm not sure.

Reply:
> static arrays that's the same size as the sprites vector.

That's tough because the sprites vector is not static, we don't know how big it is until runtime. So we'd have to overcommit to a static array that is big enough to hold what we think the upper bound amount of sprites is, which is shitty from a lot of different perspectives.

Reply:
ah, right, array size needs to be known at compile time.
perhaps just a vector then, but I don't know.